native/cocos/2d/renderer/Batcher2d.cpp

```diff
@@ -0,0 +1,258 @@
+#include <2d/renderer/Batcher2d.h>
+#include <cocos/base/TypeDef.h>
+#include <cocos/core/Root.h>
+#include <renderer/pipeline/Define.h>
+#include <scene/Pass.h>
+#include <deque>
+#include <iostream>
+
+namespace cc {
+Batcher2d::Batcher2d() : Batcher2d(nullptr) {
+    _device = Root::getInstance()->getDevice();
+}
+
+Batcher2d::Batcher2d(Root* root) : _drawBatchPool([]() { return ccnew scene::DrawBatch2D(); }, 10U) {
+    _root = root;
+    _device = Root::getInstance()->getDevice();
+}
+
+Batcher2d::~Batcher2d() {
+    _drawBatchPool.destroy();
+
+    for (auto iter = _descriptorSetCache.begin(); iter != _descriptorSetCache.end(); iter++) {
+        iter->second->destroy();
+    }
+    _descriptorSetCache.clear();
+}
+
+void Batcher2d::syncMeshBuffersToNative(std::vector<UIMeshBuffer*>&& buffers, uint32_t length) {
+    _meshBuffers = std::move(buffers);
+}
+
+UIMeshBuffer* Batcher2d::getMeshBuffer(index_t bufferId) {
+    assert(this->_meshBuffers[bufferId]);
+    return this->_meshBuffers[bufferId];
+}
+
+gfx::Device* Batcher2d::getDevice() {
+    if (_device == nullptr) {
+        _device = Root::getInstance()->getDevice();
+    }
+    return _device;
+}
+
+void Batcher2d::updateDescriptorSet() {
+    // for this._descriptorSetCache.update()
+}
+
+void Batcher2d::addRootNode(Node* node) {
+    _rootNodeArr.push_back(node);
+}
+
+//对标ts的walk
+void Batcher2d::fillBuffersAndMergeBatches() {
+    for (index_t i = 0; i < _rootNodeArr.size(); i++) {
+        walk(_rootNodeArr[i]);
+        generateBatch(_currEntity);
+    }
+    _rootNodeArr.clear();
+}
+
+void Batcher2d::walk(Node* node) {
+    static std::vector<Node*> nodeStack{1000000};
```

Comment:
Please use ccstd::vector

Reply:
And i think it is better use `ccstd::stack` instead.

Reply:
done